test/unit/integrated_function_test.C

```diff
@@ -0,0 +1,274 @@
+          libMesh::Real costheta = std::cos(theta[t]);
+
+          std::vector<std::string> functions;
+          std::vector<libMesh::Real> calc_answers;
+          std::vector<libMesh::Real> errors;
+          std::vector<bool> converged;
+
+          // trig
+          functions.push_back("sin(x)*cos(y)");
+          calc_answers.push_back( ( ( std::cos(sintheta*L-costheta*L+2.6)/(2.0*(sintheta-costheta)) ) - ( std::cos(sintheta*L+costheta*L+2.6)/(2.0*(sintheta+costheta)) ) ) - ( (std::cos(2.6)/(2.0*(sintheta-costheta))) - ( std::cos(2.6)/(2.0*(sintheta+costheta) ) ) ) );
+          errors.push_back(-1.0);
+          converged.push_back(false);
+
+          // exponential
+          functions.push_back("exp(y)/exp(x)");
+          calc_answers.push_back( (std::exp((sintheta-costheta)*L+2.6)-std::exp(2.6))/(sintheta-costheta) );
+          errors.push_back(-1.0);
+          converged.push_back(false);
+
+          GRINS::MultiphysicsSystem* system = sim.get_multiphysics_system();
+          GRINS::SharedPtr<libMesh::EquationSystems> es = sim.get_equation_system();
+
+          // build rayfire
+          GRINS::SharedPtr<GRINS::RayfireMesh> rayfire = new GRINS::RayfireMesh(origin,theta[t]);
+
+          CPPUNIT_ASSERT_EQUAL(functions.size(), calc_answers.size());
+
+          // and now the CompositeQoI to do the evalutaion
+          GRINS::CompositeQoI comp_qoi;
+
+          for (unsigned int i=0; i<functions.size(); i++)
+            {
+              GRINS::IntegratedFunction<libMesh::FunctionBase<libMesh::Real> > integ_func((unsigned int)2,new libMesh::ParsedFunction<libMesh::Real>(functions[i]),rayfire,"integrated_function");
+              comp_qoi.add_qoi(integ_func);
+            }
+
+          comp_qoi.init(input,*system);
+          system->attach_qoi(&comp_qoi);
+
+          libMesh::MeshRefinement mr(system->get_mesh());
+
+          unsigned int num_converged = 0;
+          unsigned int iter = 0;
+
+          // limit iterations to prevent excessive runtime
+          unsigned int max_iter = 10;
+
+          do
+            {
+              std::vector<libMesh::dof_id_type> elems_in_rayfire;
+              rayfire->elem_ids_in_rayfire(elems_in_rayfire);
+
+              system->assemble_qoi();
+
+              for (unsigned int i=0; i<calc_answers.size(); i++)
+                {
+                  if (!converged[i])
+                    {
+                      libMesh::Real err = std::abs( sim.get_qoi_value(i) - calc_answers[i] );
+
+                      libMesh::Real mu = std::abs(err/errors[i]);
+
+                      if (err < libMesh::TOLERANCE)
+                        {
+                          num_converged++;
+                          converged[i] = true;
+                        }
+                      else if (errors[i] == -1.0)
+                        errors[i] = err;
+                      else if (errors[i] > -1.0)
+                        {
+                          // check for linear convergence
```

Comment:
I am using this definition of linear convergence:

\mu = \frac{ |x_{n+1} - x_{n}| }{ |x_{n} - x_{n-1}| }

where 0 < \mu < 1 for convergence on each step.

The first assert I guess is more of a sanity check, since numerator and denomerator of \mu are abs(). The second assert makes sure the error of the current step is less than the error of the previous step, or \mu < 1